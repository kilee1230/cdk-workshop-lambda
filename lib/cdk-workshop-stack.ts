import { Duration, Stack, StackProps } from "aws-cdk-lib";
import { Code, Runtime, Function } from "aws-cdk-lib/aws-lambda";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";
import { Construct } from "constructs";
import { Fn } from "aws-cdk-lib";

export class CdkWorkshopStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Define an S3 bucket
    const lambdaBucket = new Bucket(this, "LambdaTestHello");

    // Deploy the Lambda code to the S3 bucket
    const lambdaCodeAsset = new BucketDeployment(
      this,
      "LambdaCodeAssetDeployment",
      {
        sources: [Source.asset("./src")],
        destinationBucket: lambdaBucket,
        exclude: ["**/*.ts"],
        extract: false,
      }
    );

    // Reference the uploaded Lambda code
    const lambdaCode = Code.fromBucket(
      lambdaCodeAsset.deployedBucket,
      Fn.select(0, lambdaCodeAsset.objectKeys)
    );

    // Define an AWS Lambda resource
    new Function(this, "HelloHandler", {
      runtime: Runtime.NODEJS_22_X,
      code: lambdaCode,
      handler: "index.handler",
      memorySize: 256,
      timeout: Duration.seconds(60),
    });
  }
}
