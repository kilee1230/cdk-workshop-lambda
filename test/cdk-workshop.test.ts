import * as cdk from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";

import { CdkWorkshopStack } from "../lib/cdk-workshop-stack";

test("S3 Bucket Created", () => {
  const app = new cdk.App();
  const stack = new CdkWorkshopStack(app, "MyTestStack");
  const template = Template.fromStack(stack);

  template.resourceCountIs("AWS::S3::Bucket", 1);
});

test("BucketDeployment Created", () => {
  const app = new cdk.App();
  const stack = new CdkWorkshopStack(app, "MyTestStack");
  const template = Template.fromStack(stack);

  template.hasResourceProperties("Custom::CDKBucketDeployment", {
    SourceBucketNames: Match.anyValue(),
    DestinationBucketName: Match.anyValue(),
    SourceObjectKeys: Match.anyValue(),
  });
});

test("Lambda Function Created", () => {
  const app = new cdk.App();
  const stack = new CdkWorkshopStack(app, "MyTestStack");
  const template = Template.fromStack(stack);

  template.hasResourceProperties("AWS::Lambda::Function", {
    Handler: "index.handler",
    Runtime: "nodejs22.x",
    MemorySize: 256,
    Timeout: 60,
  });
});

test("Snapshot Test", () => {
  const app = new cdk.App();
  const stack = new CdkWorkshopStack(app, "MyTestStack");

  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});
