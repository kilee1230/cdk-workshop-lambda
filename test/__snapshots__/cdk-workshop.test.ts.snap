// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "LambdaCodeAssetDeploymentAwsCliLayerA0D48374",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "LambdaTestHello42DF2A8D",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "LambdaTestHello42DF2A8D",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HelloHandler2E4FBA4D": {
      "DependsOn": [
        "HelloHandlerServiceRole11EF7C63",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Select": [
              0,
              {
                "Fn::Split": [
                  "/",
                  {
                    "Fn::Select": [
                      5,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "LambdaCodeAssetDeploymentCustomResource08FD9CC1",
                              "DestinationBucketArn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
          "S3Key": {
            "Fn::Select": [
              0,
              {
                "Fn::GetAtt": [
                  "LambdaCodeAssetDeploymentCustomResource08FD9CC1",
                  "SourceObjectKeys",
                ],
              },
            ],
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "HelloHandlerServiceRole11EF7C63",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Tags": [
          {
            "Key": "demo:label",
            "Value": "cdk-workshop",
          },
        ],
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "HelloHandlerServiceRole11EF7C63": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "demo:label",
            "Value": "cdk-workshop",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaCodeAssetDeploymentAwsCliLayerA0D48374": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "eafa2d6c6a55bda16b1603003f1721e2f66a024c0c31d6665ee1c5ff59b565d9.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "LambdaCodeAssetDeploymentCustomResource08FD9CC1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketArn": {
          "Fn::GetAtt": [
            "LambdaTestHello42DF2A8D",
            "Arn",
          ],
        },
        "DestinationBucketName": {
          "Ref": "LambdaTestHello42DF2A8D",
        },
        "Exclude": [
          "**/*.ts",
        ],
        "Extract": false,
        "OutputObjectKeys": true,
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": [
          "02766fc4f1301ae9b7482e3fb831c8b7d603449bad2f4bba472cc3db4c2907cc.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "LambdaTestHello42DF2A8D": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Tags": [
          {
            "Key": "aws-cdk:cr-owned:5e799c07",
            "Value": "true",
          },
          {
            "Key": "demo:label",
            "Value": "cdk-workshop",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
